src/PanoImageRenderer/vr/VRManager.ts

```diff
@@ -0,0 +1,146 @@
+import Promise from "promise-polyfill";
+import { mat4 } from "gl-matrix";
+
+const VR_DISPLAY_PRESENT_CHANGE = "vrdisplaypresentchange";
+const DEFAULT_LEFT_BOUNDS = [0, 0, 0.5, 1];
+const DEFAULT_RIGHT_BOUNDS = [0.5, 0, 0.5, 1];
+const EYES = {
+  LEFT: "left",
+  RIGHT: "right"
+};
+
+class VRManager {
+  private _vrDisplay: VRDisplay | null;
+  private _frameData: VRFrameData;
+  private _yawOffset: number;
+  private _leftBounds: number[];
+  private _rightBounds: number[];
+
+  constructor() {
+    this._frameData = new window.VRFrameData();
+    this._clear();
+  }
+
+  public get context() { return this._vrDisplay; }
+
+  public destroy = () => {
+    const vrDisplay = this._vrDisplay;
+
+    this.removeEndCallback(this.destroy);
+
+    if (vrDisplay && vrDisplay.isPresenting) {
+      vrDisplay.exitPresent();
+    }
+
+    this._clear();
+  }
+
+  public canRender() {
+    return Boolean(this._vrDisplay);
+  }
+
+  public beforeRender(gl: WebGLRenderingContext) {
+    // Render to the default backbuffer
+    gl.bindFramebuffer(gl.FRAMEBUFFER, null);
+  }
+
+  public afterRender() {
+    this._vrDisplay!.submitFrame();
+  }
+
+  public getEyeParams(gl: WebGLRenderingContext) {
+    const display = this._vrDisplay!;
+    const halfWidth = gl.drawingBufferWidth * 0.5;
+    const height = gl.drawingBufferHeight;
+    const frameData = this._frameData;
+
+    display.getFrameData(frameData);
+
+    const leftMVMatrix = frameData.leftViewMatrix;
+    const rightMVMatrix = frameData.rightViewMatrix;
+
+    mat4.rotateY(leftMVMatrix, leftMVMatrix, this._yawOffset);
+    mat4.rotateY(rightMVMatrix, rightMVMatrix, this._yawOffset);
+
+    return [
+      {
+        viewport: [0, 0, halfWidth, height],
+        mvMatrix: leftMVMatrix,
+        pMatrix: frameData.leftProjectionMatrix,
+      },
+      {
+        viewport: [halfWidth, 0, halfWidth, height],
+        mvMatrix: rightMVMatrix,
+        pMatrix: frameData.rightProjectionMatrix,
+      }
+    ];
+  }
+
+  public isPresenting() {
+    return Boolean(this._vrDisplay && this._vrDisplay.isPresenting);
+  }
+
+  public addEndCallback(callback: (evt?: Event) => any) {
+    window.addEventListener(VR_DISPLAY_PRESENT_CHANGE, callback);
+  }
+
+  public removeEndCallback(callback: (evt?: Event) => any) {
+    window.removeEventListener(VR_DISPLAY_PRESENT_CHANGE, callback);
+  }
+
+  public requestPresent(canvas: HTMLCanvasElement) {
```

Comment:
In addition to the above comment, it would be as follows.

```js
return new Promise(resolve => {
       vrDisplay.requestPresent([{source: canvas}]).then(() => {
            const leftEye = vrDisplay.getEyeParameters(EYES.LEFT);
            const rightEye = vrDisplay.getEyeParameters(EYES.RIGHT);

            canvas.width = Math.max(leftEye.renderWidth, rightEye.renderWidth) * 2;
            canvas.height = Math.max(leftEye.renderHeight, rightEye.renderHeight);

            this._setDisplay(vrDisplay);

         resolve();
    });
});
```